sql/alter.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package sql
+
+import (
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/client"
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/sql/parser"
+	"github.com/cockroachdb/cockroach/util/log"
+)
+
+// CreateIndex creates an index.
+// Privileges: ???
+func (p *planner) CreateIndex(n *parser.CreateIndex) (planNode, error) {
+	desc, err := p.getTableDesc(n.Table)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, err := desc.FindIndexByName(string(n.Name)); err == nil {
+		if n.IfNotExists {
+			// Noop.
+			return &valuesNode{}, nil
+		}
+		return nil, fmt.Errorf("index %q already exists", string(n.Name))
+	}
+
+	index := IndexDescriptor{
+		Name:        string(n.Name),
+		Unique:      n.Unique,
+		ColumnNames: n.Columns,
+	}
+	desc.Indexes = append(desc.Indexes, index)
+
+	if err := desc.AllocateIDs(); err != nil {
+		return nil, err
+	}
+
+	// `index` changed on us when we called `desc.AllocateIDs()`.
+	index = desc.Indexes[len(desc.Indexes)-1]
+
+	// Get all the rows affected.
+	// TODO(vivek): Avoid going through Select.
+	// TODO(tamird): Support partial indexes?
+	row, err := p.Select(&parser.Select{
+		Exprs: parser.SelectExprs{parser.StarSelectExpr()},
+		From:  parser.TableExprs{&parser.AliasedTableExpr{Expr: n.Table}},
+	})
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Do you have a concrete suggestion about which part of this change will not be useful later on? As far as I can tell, it will all be useful. I have some nits about this change, but we had no functionality here before and now we have something that works. Let's get that something in, perhaps with TODOs sprinkled around where we know enhancements are needed, and keep iterating.